### PR TITLE
[WIP] OrderedMap 

### LINKED ISSUE #970: `renameProperty`, `renameMember` lose order in projection
For this to be true #948, we need to resolve the bug where a property that gets renamed gets moved to the end of the property list/map(Because it is removed then added)

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #970

Work in progress of making a custom map that support renaming keys without loosing order. This would unlock the ability to keep properties, memebers in order in projections

## Benchmark
Does seem to take a bit of a hit, my guess is the` #value` is much less efficient with object keys than string keys

```
Time Map 329
Time OrderedMap 1222
```

code used
```ts
function time(name: string, callback: () => void) {
      const start = new Date().getTime();
      callback();
      const end = new Date().getTime();
      console.log(`Time ${name}`, end - start);
    }
    function runBenchmark(map: Map<string, string>) {
      for (let i = 0; i < 10000; i++) {
        const key = i.toString();
        map.set(key, key);
      }
    }
    time("Map", () => {
      for (let i = 0; i < 1000; i++) {
        runBenchmark(new Map<string, string>());
      }
    });
    time("OrderedMap", () => {
      for (let i = 0; i < 1000; i++) {
        runBenchmark(new OrderedMap<string>());
      }
    });
```

Getting similar numbers if changing the number of items to `100` but runnning many more times. So number of items in the map doesn't affect much(as it should)

### CODE REVIEW

**Comment by nguerrera** [comment]
I'd like to see performance measurement on this. I don't know if it will show up overall, but it seems that this would basically double the cost of most map operations.

File: packages/compiler/core/util.ts [hunk 1]
```diff
@@ -322,3 +322,95 @@ class CaseInsensitiveMap<T> extends Map<string, T> {
...
+
+    this.#values.set(keyItem, value);
+    return this;
+  }
+
+  get size() {
+    return this.#values.size;
+  }
+
+  entries(): IterableIterator<[string, V]> {
```
Comment (bterlson) [inline packages/compiler/core/util.ts:384]: These should probably be implemented as generators so we don't create the intermediate array in memory just to peel off an iterator for its items.

Comment (timotheeguerin) [reply packages/compiler/core/util.ts:384]: generator or iterator? I was under the impression that generators have pretty bad performance

Comment (bterlson) [reply packages/compiler/core/util.ts:384]: Generator, hand-rolling an iterator would require some pretty intense transforms that would definitely be slower. I don't think generators are that slow anymore, but regardless I would say that for a general purpose data structure the allocation of the intermediate array is the worse evil.

File: packages/compiler/core/util.ts [hunk 2]
```diff
@@ -322,3 +322,102 @@ class CaseInsensitiveMap<T> extends Map<string, T> {
     return super.delete(key.toUpperCase());
   }
 }
+
+interface OrderedMapKey {
+  key: string;
+}
+
+export class OrderedMap<V> implements Map<string, V> {
```
Comment (timotheeguerin) [inline packages/compiler/core/util.ts:330]: Could technically also have this take the `K` template argument. The key doesn't need to be a string for this to work. Don't think we'll have a use for it but it should work.

Comment (bterlson) [reply packages/compiler/core/util.ts:330]: I think using the same type signature as map will be a win just from a readability standpoint.

File: packages/compiler/core/util.ts [hunk 3]
```diff
@@ -322,3 +322,102 @@ class CaseInsensitiveMap<T> extends Map<string, T> {
     return super.delete(key.toUpperCase());
   }
 }
+
+interface OrderedMapKey<K> {
+  key: K;
+}
+
+export class OrderedMap<K, V> implements Map<K, V> {
```
Comment (timotheeguerin) [inline packages/compiler/core/util.ts:330]: Name is also maybe not the best, as the additional functionality it provide over `Map` is not ordering but renaming a key while keeping order.

**Comment by kevinswiber** [comment]
> Does seem to take a bit of a hit, my guess is the #value is much less efficient with object keys than string keys

The benchmark is doing a sequence of `Map#get`, `Map#set`, `Map#set` for each `OrderedMap#set` operation.  I wouldn't suspect string vs. object keys are having a big impact here.  You could try turning `values` into a `WeakMap<OrderedMapKey<K>, V>`, but even then, I think any perf gain on `set` would have a trade-off of worse perf on iteration.

**Comment by nguerrera** [comment]
I've tried a lot of different approaches with different theoretical tradeoffs but nothing is materially better perf-wise that I've found. I will now try to measure the impact end-to-end on a small compilation rather than microbenchmark the map itself to assess if we can just use this.

**Pull request closed without merging**